apps/bmd/src/AppRoot.tsx

```diff
@@ -141,898 +147,1091 @@ export const activationStorageKey = 'activation'
+        votes: {
+          ...state.votes,
+          [action.contestId]: action.vote,
+        },
+      }
+    }
+    case 'forceSaveVoteFlag':
+      return {
+        ...state,
+        forceSaveVoteFlag: action.value,
+      }
+    case 'resetBallot':
+      return {
+        ...state,
+        ...initialCardState,
+        ...initialVoterState,
+      }
+    case 'setUserSettings':
+      /* istanbul ignore next */
+      if (Object.keys(action.userSettings).join(',') !== 'textSize') {
+        throw new Error('unknown userSetting key')
+      }
+      if (action.userSettings.textSize === state.userSettings.textSize) {
+        return state
+      }
+      return {
+        ...state,
+        userSettings: {
+          ...state.userSettings,
+          ...action.userSettings,
+        },
+      }
+    case 'updateAppPrecinctId':
+      return {
+        ...state,
+        ...resetTally,
+        appPrecinctId: action.appPrecinctId,
+      }
+    case 'enableLiveMode':
+      return {
+        ...state,
+        ...resetTally,
+        isLiveMode: true,
+        isPollsOpen: initialAppState.isPollsOpen,
+      }
+    case 'toggleLiveMode':
+      return {
+        ...state,
+        ...resetTally,
+        isLiveMode: !state.isLiveMode,
+        isPollsOpen: initialAppState.isPollsOpen,
+      }
+    case 'togglePollsOpen':
+      return {
+        ...state,
+        isPollsOpen: !state.isPollsOpen,
+      }
+    case 'updateTally': {
+      const { electionDefinition, tally, votes } = state
+      return {
+        ...state,
+        ballotsPrintedCount: state.ballotsPrintedCount + 1,
+        tally: calculateTally({
+          election: electionDefinition!.election,
+          tally,
+          votes: votes!,
+        }),
+      }
+    }
+    case 'updateElectionDefinition':
+      return {
+        ...state,
+        electionDefinition: action.electionDefinition,
+      }
+    case 'startWritingLongValue':
+      return {
+        ...state,
+        writingVoteToCard: true,
+        forceSaveVoteFlag: false,
+        lastVoteSaveToCardAt: Date.now(),
+      }
+    case 'finishWritingLongValue':
+      return {
+        ...state,
+        writingVoteToCard: false,
+      }
+    case 'updateHardwareState':
+      return {
+        ...state,
+        ...action.hardwareState,
+      }
+    case 'initializeAppState':
+      return {
+        ...state,
+        ...action.appState,
+      }
+    case 'updateLastCardDataString': {
+      return {
+        ...state,
+        lastCardDataString: action.currentCardDataString,
+      }
+    }
   }
+}
 
-  public enableLiveMode = (): void => {
-    this.setState(
-      { isLiveMode: true, isPollsOpen: this.initialUserState.isPollsOpen },
-      this.resetTally
-    )
-  }
+const AppRoot: React.FC<Props> = ({
+  card,
+  hardware,
+  history,
+  machineConfig: machineConfigProvider,
+  printer,
+  storage,
+}) => {
+  const [appState, dispatchState] = useReducer(appReducer, initialAppState)
```

Comment:
A question on terminology… do we want `dispatch` as per the [examples in the docs](https://reactjs.org/docs/hooks-reference.html#usereducer)? Or is there a compelling case for `dispatchState`?

Reply:
If we have more than one `useReducer`, they should definitely have unique names. It seemed that `dispatch` was too generic, but perhaps it should mirror the state name, so I'm gonna change it to `dispatchAppState` for now. Happy to change it later if we want to.